core/src/main/java/org/elasticsearch/cluster/routing/allocation/AllocationService.java

```diff
@@ -239,12 +239,44 @@ static MetaData updateMetaDataWithRoutingTable(MetaData oldMetaData, RoutingTabl
             return new RoutingAllocation.Result(false, clusterState.routingTable(), clusterState.metaData());
         }
         gatewayAllocator.applyFailedShards(allocation);
+        // elect primaries *before* allocating unassigned, so backups of primaries that failed
+        // will be moved to primary state and not wait for primaries to be allocated and recovered (*from gateway*)
+        electPrimariesAndUnassignedDanglingReplicas(allocation);
+
         reroute(allocation);
         String failedShardsAsString = firstListElementsToCommaDelimitedString(failedShards, s -> s.shard.shardId().toString());
         return buildResultAndLogHealthChange(allocation, "shards failed [" + failedShardsAsString + "] ...");
     }
 
     /**
+     * unassigned an shards that are associated with nodes that are no longer part of the cluster, potentially promoting replicas
+     * if needed.
+     */
+    public RoutingAllocation.Result deassociateDeadNodes(ClusterState clusterState, boolean reroute, String reason) {
```

Comment:
how about adding an overloaded version of the method where reroute is true (like we have for startedShards / failedShards)?

alternatively, we could make it even more explicit when reroute is not called, i.e. have `deassociateDeadNodes` always do the reroute and add a method `deassociateDeadNodesWithoutReroute` for the rare cases where we don't reroute.